### PR TITLE
Debug render

### PR DESCRIPTION
**Pull request by jesstelford**
## Debug Renderer

A debug renderer for the `<World>` component, overlaid on your scene:

demo

### Motivation

To easily debug matter-js world objects. I initially leveraged [`Matter.Render`](https://github.com/liabru/matter-js/wiki/Rendering#using-matterrender), and figured others might find it useful.

### Implementation

Added a new prop to `<World>`:

```javascript
    debug: PropTypes.shape({
      offset: PropTypes.shape({
        x: PropTypes.number,
        y: PropTypes.number,
      }),
      background: PropTypes.string,
    }),
```

By setting this prop, a `<canvas>` will be overlaid on the screen, showing wireframes of physics objects.

Unlike regular react props, defaults are deeply merged into the `debug` prop, allowing partial setting of the object.

### Notes

* Relies on #21 _"Surface display width and height in stage context"_
* Comments have been left inline describing specific features.
* Only works for web (I'm not familiar with how to render to canvas in React Native)

### CODE REVIEW

File: demo/game/index.js [hunk 1]
```diff
@@ -49,6 +52,12 @@ export default class Game extends Component {
     Matter.World.addBody(engine.world, ground);
     Matter.World.addBody(engine.world, leftWall);
     Matter.World.addBody(engine.world, rightWall);
+
+    Matter.Events.on(engine, 'afterUpdate', this.update);
+
+    const unsubscribeFromUpdate = () => {
+      Matter.Events.off(engine, 'afterUpdate', this.update);
+    }
```
Comment (jesstelford) [inline demo/game/index.js]: Setting up events to listen to keypresses on every frame.

File: demo/game/index.js [hunk 2]
```diff
@@ -80,18 +104,25 @@ export default class Game extends Component {
       fade: false,
     });
 
+    this.stageXUIUnsubscribe = GameStore.onStageXChange((stageX) => {
+      this.forceUpdate();
+    });
```
Comment (jesstelford) [inline demo/game/index.js]: As the stage is scrolled, we want to also move our debug canvas, so we force an update on change.

File: demo/game/index.js [hunk 3]
```diff
@@ -100,6 +131,13 @@ export default class Game extends Component {
         <Stage style={{ background: '#3a9bdc' }}>
           <World
             onInit={this.physicsInit}
+            debug={this.state.debug && {
+              offset: {
+                x: -GameStore.stageX,
```
Comment (jesstelford) [inline demo/game/index.js]: The "stage" is manually scrolled, so too must our debug canvas.

File: src/components/world.js [hunk 4]
```diff
@@ -42,6 +51,94 @@ export default class World extends Component {
...
+  };
+
+  stopDebugRendering = () => {
+    if (this._render) {
+      Render.stop(this._render);
+      delete this._render;
+    }
+  };
+
+  getDebugProps = () => {
```
Comment (jesstelford) [inline src/components/world.js]: Performs a deep merge of the debug props.

File: src/components/world.js [hunk 5]
```diff
@@ -42,6 +51,94 @@ export default class World extends Component {
...
+      options: {
+        wireframeBackground: background,
+        width: renderWidth,
+        height: renderHeight,
+      },
+    });
+
+    // Setting this as part of `.create` crashes Chrome during a deep clone. :/
+    // My guess: a circular reference
+    this._render.engine = this.engine;
```
Comment (jesstelford) [inline src/components/world.js]: This was a head scratcher for sure! Other than fixing the implementation inside `matter-js`, I couldn't find any other way to stop Chrome from crashing during the deep merge above.

File: src/components/world.js [hunk 6]
```diff
@@ -91,9 +220,26 @@ export default class World extends Component {
...
 
+    const { renderWidth, renderHeight, scale } = this.context;
+
+    let debugRenderTarget = false;
+
+    if (this.props.debug) {
+      debugRenderTarget = (
+        <canvas
+          key="debug-render-target"
+          style={{position: 'relative'}}
```
Comment (jesstelford) [inline src/components/world.js]: So it sits on top of the other relatively positioned elements in its stacking context.

**Comment by kenwheeler** [comment]
THIS IS BANANAS

**Pull request merged**